File: backend/app/schemas/round.py
from pydantic import BaseModel, ConfigDict
from typing import Optional, List
from datetime import datetime
from app.models.round import RoundStatus
from app.schemas.song import SongCreate, SongResponse


class RoundBase(BaseModel):
    theme: str
    description: Optional[str] = None


class RoundCreate(RoundBase):
    league_id: int
    order: Optional[int] = None  # Optional, defaults to last position


class RoundUpdate(BaseModel):
    theme: Optional[str] = None
    description: Optional[str] = None
    # Note: deadlines are auto-calculated, not editable
    # Status changes should use dedicated endpoints (start, complete)


class RoundResponse(RoundBase):
    id: int
    league_id: int
    league_name: Optional[str] = None
    order: int
    status: RoundStatus

    # Timestamps (nullable until round starts)
    started_at: Optional[datetime] = None
    submission_deadline: Optional[datetime] = None
    voting_started_at: Optional[datetime] = None
    voting_deadline: Optional[datetime] = None
    completed_at: Optional[datetime] = None

    created_at: datetime
    submission_count: Optional[int] = None
    user_has_submitted: Optional[bool] = None
    user_has_voted: Optional[bool] = None
    is_admin: Optional[bool] = None
    songs_per_round: Optional[int] = None  # From league settings

    model_config = ConfigDict(from_attributes=True)


class SubmissionCreate(BaseModel):
    round_id: int
    songs: List[SongCreate]


class SubmissionResponse(BaseModel):
    id: int
    round_id: int
    user_id: int
    submitted_at: datetime
    songs: List[SongResponse] = []
    user_name: Optional[str] = None  # Only shown after voting

    model_config = ConfigDict(from_attributes=True)


class RoundDetailResponse(RoundResponse):
    submissions: List[SubmissionResponse] = []
    
    model_config = ConfigDict(from_attributes=True)

File: frontend/src/types/round.ts
import { Song, SongInput } from './song';

export enum RoundStatus {
  PENDING = 'pending',
  ACTIVE = 'active',
  COMPLETED = 'completed'
}

export interface Round {
  id: number;
  league_id: number;
  league_name?: string | null;
  theme: string;
  description: string | null;
  order: number;
  status: RoundStatus;

  // Timestamps (null until round starts)
  started_at: string | null;
  submission_deadline: string | null;
  voting_started_at: string | null;
  voting_deadline: string | null;
  completed_at: string | null;

  created_at: string;
  submission_count?: number;
  user_has_submitted?: boolean;
  user_has_voted?: boolean;
  is_admin?: boolean;
  songs_per_round?: number;  // From league settings
}

export interface Submission {
  id: number;
  round_id: number;
  user_id: number;
  submitted_at: string;
  songs: Song[];
  user_name?: string | null;
}

export interface RoundDetail extends Round {
  submissions: Submission[];
}

export interface RoundCreate {
  league_id: number;
  theme: string;
  description?: string;
  order?: number;  // Optional, defaults to last position
}

export interface RoundUpdate {
  theme?: string;
  description?: string;
  // Note: deadlines are auto-calculated, not editable
}

export interface RoundReorderItem {
  id: number;
  order: number;
}

export interface SubmissionCreate {
  round_id: number;
  songs: SongInput[];
}
